From 7349b7c5aa58b937394f5a44844238ff38e2f80d Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Thu, 16 Apr 2026 08:34:38 +0100
Subject: [PATCH] daemon/listfs.ml: Ignore CHS geometry error from parted
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

parted has an annoying bug where it fails to parse Sun partition
tables (still used by Veritas).  It tries to check the CHS geometry
stored in the header matches the physical CHS geometry, which is a
meaningless test.

In function is_partition_can_hold_filesystem we are only interested in
MBR and GPT partition types, so catch and ignore this specific parted
error.

I tested this by adding a Sun disk as a second disk to a Fedora
guest and doing inspection.  Previously the operation would
fail:

  $ guestfish --ro -a /var/tmp/fedora-41.img -a /var/tmp/dump.img -i
  libguestfs: error: inspect_os: parted exited with status 1: Warning:
  The disk CHS geometry (205603,255,2) reported by the operating
  system does not match the geometry stored on the disk label
  (64887,16,101).

After this change it succeeds:

  $ guestfish --ro -a /var/tmp/fedora-41.img -a /var/tmp/dump.img -i

  Welcome to guestfish, the guest filesystem shell for
  editing virtual machine filesystems and disk images.

  Type: ‘help’ for help on commands
        ‘man’ to read the manual
        ‘quit’ to quit the shell

  Operating system: Fedora Linux 41 (Forty One)
  /dev/sda3 mounted on /
  /dev/sda2 mounted on /boot

Fixes: https://redhat.atlassian.net/browse/RHEL-165220
(cherry picked from commit 8d83bf2bcf31c5206b6deaa5e993009a85ff174f)
---
 daemon/listfs.ml | 44 ++++++++++++++++++++++++++++----------------
 1 file changed, 28 insertions(+), 16 deletions(-)

diff --git a/daemon/listfs.ml b/daemon/listfs.ml
index 067314c47..8ca5ca8a8 100644
--- a/daemon/listfs.ml
+++ b/daemon/listfs.ml
@@ -118,30 +118,42 @@ and is_partition_can_hold_filesystem partition =
   let device = Devsparts.part_to_dev partition
   and partnum = Devsparts.part_to_partnum partition in
 
-  match Parted.part_get_parttype device with
-  | "msdos" ->
-     if Parted.part_get_mbr_part_type device partnum = "extended" then
-       false
-     else if partnum = 1 && Utils.has_bogus_mbr device then
-       true
-     else
-       true
+  try
+    match Parted.part_get_parttype device with
+    | "msdos" ->
+       if Parted.part_get_mbr_part_type device partnum = "extended" then
+         false
+       else if partnum = 1 && Utils.has_bogus_mbr device then
+         true
+       else
+         true
 
-  | "gpt" ->
-     let gpt_type = Sfdisk.part_get_gpt_type device partnum in
-     (match gpt_type with
-      (* Windows Logical Disk Manager metadata partition. *)
-      | "5808C8AA-7E8F-42E0-85D2-E1E90434CFB3"
+    | "gpt" ->
+       let gpt_type = Sfdisk.part_get_gpt_type device partnum in
+       (match gpt_type with
+        (* Windows Logical Disk Manager metadata partition. *)
+        | "5808C8AA-7E8F-42E0-85D2-E1E90434CFB3"
         (* Windows Logical Disk Manager data partition. *)
         | "AF9B60A0-1431-4F62-BC68-3311714A69AD"
         (* Microsoft Reserved Partition. *)
         | "E3C9E316-0B5C-4DB8-817D-F92DF00215AE"
         (* Windows Snapshot Partition. *)
         | "CADDEBF1-4400-4DE8-B103-12117DCF3CCF" -> false
-      | _ -> true
-     )
+        | _ -> true
+       )
 
-  | _ -> (* unknown or other *)
+    | _ -> (* unknown or other *)
+       true
+
+  with
+  | Failure msg when String.find msg "CHS geometry" >= 0 ->
+     (* Parted has poor handling of "sun" partition types, always
+      * issuing a warning because the CHS doesn't match the physical
+      * geometry.  Ignore this as we don't care about it in this
+      * function (RHEL-165220).
+      *)
+     eprintf "is_partition_can_hold_filesystem: \
+              ignoring warning from parted: %s\n%!" msg;
      true
 
 (* Use vfs-type to check for a filesystem of some sort of [device].
-- 
2.47.3

